unison-src/transcripts/delete.output.md

```diff
@@ -182,10 +188,6 @@ type Foo = Foo Boolean
     5. ┌ Foo.Foo#d97e0jhkmd#0 : Nat -> Foo
     6. └ Foo.Foo#gq9inhvg9h#0 : Boolean -> b.Foo
   
-  Added definitions:
-  
-    7. foo : Nat
-  
```

Comment:
This old output was incorrect: `foo` should have been deleted in the `a` namespace, so it shouldn't be resurfacing here.

Reply:
Seems like a merge error.  We had `.a.foo = 1` and `.b.foo = 2`, then we merged `.b` into `.a`, and deleted `.a.foo`.  So we have no `.a.foo` but we still have `.b.foo`.  Then we merged `.b` into `.a` again, and had previously ended up  with `.a.foo = 2` (presumably) again.

But why did it change now?

Since the version of `.b` that contained `foo` was already part of the history of `.a`, it seems like it shouldn't have been added again.  I wouldn't have been surprised if this changed with some `lca` or `merge` thing, but I still can't really imagine the issue and don't get what it has to do with this PR.

Reply:
I think I get it: old implementation would have a chain of history for each definition added, and moreover the definitions in the chain would be in an arbitrary order (or maybe alphabetical). This then affects LCA calculation and the merge result. I’ll see if I can confirm this hypothesis.

Reply:
Interesting, okay.

Reply:
Okay, got to the bottom of this and it's not what I thought. The difference in behavior is because the new code leaves around an empty sub-branch which remembers the causal history that led to that sub-branch being empty. The old code was over-eager about deleting sub-branches that became empty. Here's [transcript output, showing the history](https://gist.github.com/pchiusano/9edb8e546149bce49ef1574cdf195577)

The key is that in the `a` namespace in this PR, history after the delete looks like what you'd expect:

```
.a> history

  Note: The most recent namespace hash is immediately below this
        message.
  
  ⊙ #gcq0ml6diu
  
    - Deletes:
    
      foo#0ja foo#jk1
  
  This segment of history starts with a merge. Use
  `history #som3n4m3space` to view history starting from a given
  namespace hash.
  
  #biv51qeect
  #u8lsnrulgl
  ⑂
  ⊙ #l24km9hmve
```

This reflects that a merge occurred and then `foo` (both versions) were then deleted. Whereas in `master`, this same portion of the transcript looks like:

```
.a> history

  ☝️  The namespace .a is empty.
```

As a result, when doing the subsequent merge of `b` into `a`, `master` has "forgotten" that at an earlier point in `a` history, `foo` was deleted. So it treats the `foo` as a new definition and adds it.

So this branch has the correct behavior, however I was curious what made the old code so eager to delete empty branches. Old code was defined in terms of:

https://github.com/unisonweb/unison/blob/b5cae909fc15639d8f6e86fc86b83141700a2675/parser-typechecker/src/Unison/Codebase/Branch.hs#L689-L696

Which as you can see just deletes the key if you ever go to insert something whose `head` is `Branch.empty0`.

So basically, this PR just happens to work due better in this instance due to a coincidence of the implementation. I'd like to understand better when it is safe to delete empty sub-branches. This seems to be a case where being overeager to delete them leads to worse merge results. I suspect you should just never delete empty sub-branches, unless the user has literally typed `delete.namespace foo.bar`. You shouldn't just delete all the history just because the namespace is currently empty... that would be a little like deleting a directory automatically once you delete the last file.

Reply:
Doh, good work.  Not sold on the relevance of the directory analogy, but agree we shouldn't be deleting info needed to merge correctly :)